### PR TITLE
Unify code path for MULTI and AUTO CTPUT hint

### PR DESCRIPTION
**Pull request by wgzintel**
### Details:
- *[MULTI] pass through to AUTO with CTPUT hint*
- *After this change*
--  MULTI doesn't support setting infer request via CPU(4),GPU(8). 
--  MULTI doesn't support CompiledModel::set_property() and ExecutableNetwork::GetConfig(). TODO: CVS-104942, CVS-106972

### Tickets:
 - *CVS-104939*

### CODE REVIEW

File: src/inference/src/dev/core_impl.cpp
Comment (wangleis) [inline src/inference/src/dev/core_impl.cpp:132]: Move this logic into auto plugin code

File: src/inference/src/dev/core_impl.cpp
Comment (wangleis) [inline src/inference/src/dev/core_impl.cpp:183]: Please add log

**Comment by wangleis** [comment]
Need separated plugin setting for AUTO and MULTI

**Comment by wenjiew** [comment]
Is it valid for this PR still in "Draft" mode?

File: src/inference/tests/functional/caching_test.cpp
Comment (songbell) [inline src/inference/tests/functional/caching_test.cpp:2095]: try  do not check device num when model path is not empty in AUTO ctput mode in auto plugin.cpp

File: src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:74]: why not? 
plugins must return their properties

File: src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp:200]: we need to fix this test and remove condition for auto / multi

File: src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp:67]: why?

File: src/inference/tests/functional/caching_test.cpp
Comment (ilya-lavrenov) [inline src/inference/tests/functional/caching_test.cpp:2226]: remove all skips here

**Review by ilya-lavrenov — CHANGES_REQUESTED** [review]

File: src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp
Comment (songbell) [reply src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp]: this can be fixed by integrating the bind PR, multi + bind, currently this part of code is missing @wgzintel

File: src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp:216]: why do we need AUTO tests in shared common tests?
They are by definition should be located in AUTO plugin folder.

Shared folder are allowed to contain only generic tests which are valid for all plugins

File: src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:46]: the same - move auto specific tests to AUTO plugin folder

File: src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp
Comment (ilya-lavrenov) [inline src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:305]: what is auto plugin specific here?

Comment (wangleis) [reply src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:305]: So far AUTO ExecutableNetwork only has RO properties. 

So AUTO ExecutableNetwork does not support GetConfig() yet.

File: src/plugins/auto/auto_schedule.cpp
Comment (riverlijunjie) [inline src/plugins/auto/auto_schedule.cpp:699]: Why `_pCTPUTLoadContext` branch add lock, but not added here?

File: src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp
Comment (ilya-lavrenov) [reply src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:305]: The similar behavior we have for other HW plugins

File: src/plugins/auto/auto_schedule.cpp
Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp]: you are right, lock is needed for fallback case

File: src/plugins/auto/auto_executable_network.cpp
Comment (songbell) [inline src/plugins/auto/auto_executable_network.cpp:74]: please optimize all these, looks too redundant. also, no matter newAPI or not, the perfmode property can work with string<->enum well.

File: src/plugins/auto/auto_schedule.cpp
Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp]: @WeldonWangwang

File: src/inference/tests/functional/caching_test.cpp
Comment (iefode) [inline src/inference/tests/functional/caching_test.cpp:2214]: REMOVE gtest skip from here! Use `skip_tests_config` instead of this construction

File: src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp
Comment (iefode) [reply src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp:67]: We don't use hardcoded `GTEST_SKIP`!

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp
Comment (iefode) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:1322]: minor: does not

**Review by iefode — CHANGES_REQUESTED** [review]

File: src/bindings/c/tests/ov_compiled_model_test.cpp
Comment (wgzintel) [reply src/bindings/c/tests/ov_compiled_model_test.cpp]: Updated

File: src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp
Comment (wgzintel) [reply src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp]: Updated

File: src/plugins/auto/auto_schedule.cpp
Comment (WeldonWangwang) [reply src/plugins/auto/auto_schedule.cpp]: Because _autoSContext->_devicePriorities[i] will be modified when fallback in _pCTPUTLoadContext, so add lock to _devicePriorities for _pCTPUTLoadContext.

File: src/plugins/auto/auto_schedule.cpp
Comment (wangleis) [inline src/plugins/auto/auto_schedule.cpp:319]: what is this `{`  used for?

File: src/tests/unit/auto/auto_default_perf_hint_test.cpp
Comment (wangleis) [inline src/tests/unit/auto/auto_default_perf_hint_test.cpp:221]: please update test case instead of removing test case

File: src/plugins/auto/auto_schedule.cpp
Comment (peterchen-intel) [reply src/plugins/auto/auto_schedule.cpp:332]: To limit the scope fo https://en.cppreference.com/w/cpp/thread/lock_guard

File: src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp
Comment (wgzintel) [reply src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp:94]: Updated

File: src/tests/unit/auto/auto_default_perf_hint_test.cpp
Comment (wgzintel) [reply src/tests/unit/auto/auto_default_perf_hint_test.cpp:221]: Updated

File: src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp
Comment (peterchen-intel) [reply src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp:62]: Confirmed that they not supported by AUTO ExecutaleNetwork.GetMetric

Comment (wgzintel) [reply src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp:62]: AUTO getMetric() does not support ov::hint::num_requests, ov::hint::allow_auto_batching and ov::enable_profiling

File: src/tests/unit/auto/auto_default_perf_hint_test.cpp
Comment (wgzintel) [reply src/tests/unit/auto/auto_default_perf_hint_test.cpp:221]: Updated

File: src/plugins/auto/auto_schedule.cpp
Comment (songbell) [inline src/plugins/auto/auto_schedule.cpp:247]: please revert this change, we don't need a member varible here

File: src/plugins/auto/auto_schedule.cpp
Comment (songbell) [reply src/plugins/auto/auto_schedule.cpp:319]: same question here, why?

File: src/plugins/auto/auto_executable_network.cpp
Comment (songbell) [inline src/plugins/auto/auto_executable_network.cpp:21]: [0] might not load success here

File: src/tests/functional/plugin/gpu/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp
Comment (songbell) [inline src/tests/functional/plugin/gpu/shared_tests_instances/behavior/ov_plugin/properties_tests.cpp:147]: 1) here delete allow_auto_batch, but shouldn' this go together with PR removal of batch keys from auto list?
2) enable_profiling is removed , why?

**Review by yangwang201911 — APPROVED** [review]

**Review by WeldonWangwang — APPROVED** [review]

**Review by songbell — APPROVED** [review]

**Review by wangleis — APPROVED** [review]

File: src/plugins/auto/auto_executable_network.cpp
Comment (songbell) [inline src/plugins/auto/auto_executable_network.cpp:82]: minor: remove line 90 lock

Comment (wgzintel) [reply src/plugins/auto/auto_executable_network.cpp:82]: Updated

File: src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/core_integration.cpp
Comment (iefode) [inline src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/core_integration.cpp:344]: Why do you remove `MULTI`? Could you please explain?

File: src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp
Comment (iefode) [inline src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp:111]: What if we don't have `GPU.1`?? 
GPU.0 is default

File: src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp
Comment (iefode) [inline src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:75]: Why do you remove `ASSERT_NO_THROW`? I suppose we should do not change

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp
Comment (iefode) [inline src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:1385]: What is the hardcode in the shared part???

File: src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp
Comment (iefode) [inline src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp:204]: The same: why do you change such construction?

File: src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/core_integration.cpp
Comment (peterchen-intel) [reply src/plugins/intel_cpu/tests/functional/shared_tests_instances/behavior/ov_plugin/core_integration.cpp:344]: @iefode as the review in Arch Forum(CVS-104405), the decision is to align the behaviors of MULTI to AUTO cumulative_throuhgput hint in 23.0 and add some of the behaviors back in 23.1/2, so this test case is not suitable for MULTI now.

**Review by ilya-lavrenov — APPROVED** [review]

File: src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp
Comment (peterchen-intel) [reply src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp:204]: Adopted from the implementation of CompiledModel::get_property(), if GetConfig fails and then try GetMetric. ASSERT_NO_THROW on GetMetric

File: src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp
Comment (peterchen-intel) [reply src/tests/functional/plugin/shared/include/behavior/ov_plugin/core_integration.hpp:1385]: @ilya-lavrenov As we agreed before, MULTI is the only one which is with THROUGHPUT as default hint, actually AUTO is the only plugin which implemented this requirement.  If your recent PRs have implemented the requirement CVS-98248 (latency as default hint) for other plugins?

File: src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp
Comment (peterchen-intel) [reply src/tests/functional/plugin/gpu/shared_tests_instances/multi/gpu_remote_blob_tests.cpp:111]: Add the comments to the code to explain the logic here.

File: src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp
Comment (peterchen-intel) [reply src/tests/functional/plugin/shared/include/behavior/executable_network/exec_network_base.hpp:75]: If AUTO/MULTI can't support this test case, skip it in skip file and with a JIRA ticket to follow up.

File: src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp
Comment (peterchen-intel) [reply src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp:204]: If AUTO/MULTI can't support this test case, skip it in skip file and with a JIRA ticket to follow up.

Comment (peterchen-intel) [reply src/tests/functional/plugin/shared/src/behavior/plugin/configuration_tests.cpp:204]: TODO: CVS-104942

**Review by iefode — APPROVED** [review]

**Comment by peterchen-intel** [comment]
Waiting for the approval from CCB

**Comment by peterchen-intel** [comment]
Approved.

**Pull request merged**
